### PR TITLE
Fixes #21154 - copy srpms into content views

### PR DESCRIPTION
**Pull request by jlsherrill**
in addition to maintain the optimizations for
foreman proxy sync, we must index srpms into our db.
Also to improve the experience, added srpm counts
to repository lists and content view version lists

### CODE REVIEW

**Comment by jlsherrill** [comment]
requires https://github.com/Katello/runcible/pull/198

**Comment by jlsherrill** [comment]
[test]

**Comment by jlsherrill** [comment]
[test]

File: app/models/katello/srpm.rb [hunk 1]
```diff
@@ -0,0 +1,40 @@
+module Katello
+  class Srpm < Katello::Model
+    include Concerns::PulpDatabaseUnit
+
+    CONTENT_TYPE = Pulp::Rpm::CONTENT_TYPE
+
+    has_many :repositories, :through => :repository_srpms, :class_name => "Katello::Repository"
+    has_many :repository_srpms, :class_name => "Katello::RepositorySrpm", :dependent => :destroy, :inverse_of => :rpm
```
Comment (parthaa) [inline app/models/katello/srpm.rb]: ```:inverse_of => :srpm```

**Comment by jlsherrill** [comment]
to validate:
1) sync a srpm repo
2) add it to a content view
3) publish the content view
4) verify the repo in the content view contains the srpms on the filesystem
5) I would also verify that check_matching_content works, where by the repo's repomd.xml file is not updated with a new timestamp when promoting a version with no changes (but is updated when promoting a version with changes).  look in the repomd.xml for its 'revision' field

File: app/models/katello/repository_srpm.rb [hunk 2]
```diff
@@ -0,0 +1,7 @@
+module Katello
+  class RepositorySrpm < Katello::Model
+    # Do not use active record callbacks in this join model.  Direct INSERTs and DELETEs are done
+    belongs_to :repository, :inverse_of => :repository_srpms, :class_name => 'Katello::Repository'
+    belongs_to :srpm, :inverse_of => :repository_srpms, :class_name => 'Katello::Srpm'
```
Comment (parthaa) [inline app/models/katello/repository_srpm.rb]: specify the inverses here so that the deletion can happen cleanly

**Comment by parthaa** [comment]
@jlsherrill might not be in scope for this PR, but I think we should add Srpm count in repository-info.html also. You just show it on product  -> repositories, but the same count is not visible on the repo details page.

File: engines/bastion_katello/app/assets/javascripts/bastion_katello/products/details/repositories/views/product-repositories.html [hunk 3]
```diff
@@ -116,6 +116,12 @@
               </div>
 
               <div>
+                <span translate ng-show="repository.content_counts.srpm > 0">
+                  {{ repository.content_counts.srpm }} Source Rpms
```
Comment (parthaa) [inline engines/bastion_katello/app/assets/javascripts/bastion_katello/products/details/repositories/views/product-repositories.html]: can you keep the case of rpms consistent I prefer ```Source RPMs``` like you have for CVV

**Comment by parthaa** [comment]
Sounds like if you upload source rpms it does not update the count.

**Comment by jlsherrill** [comment]
@parthaa updated

**Comment by parthaa** [comment]
> Sounds like if you upload source rpms it does not update the count.

I created http://projects.theforeman.org/issues/21412  . Hopefully we can address this issue there..

**Comment by parthaa** [comment]
APJ

**Comment by parthaa** [comment]
Not sure if this is an indexing issue or something else...
- Sync a source rpm repo  (I did fedora, on demand and cancelled midway but pulp had enough data. I am sure it reproducible with a simpler repo)
- Delete that repo via UI
- Now in the rails console try this test
```
> ::Katello::RepositorySrpm.count
=> 0  # Good! no repos with srpms
> ::Katello::Srpm.import_all
=> ....... 10 rest calls later
=> # it got me 3K packages to index
> ::Katello::Srpm.count
=> 3842 # I guess I needed to run clean up orphans
```
- Run the the clean up 
```
$ rake katello:delete_orphan_content
.....
```
- get back to rails console
```
> ::Katello::Srpm.import_all
=> ....... 1 rest call later which tells me there are no srpms

> ::Katello::Srpm.count
=> 3842
# Tells me SRPM count is not getting changed even if upstream pulp says there no source rpms
```
I also know pulp srpm is clear here because I checked in mongo
```
$ mongo pulp_database
> db.units_rpm.count()
8636

> db.units_srpm.count()
0
```
So if the indexing operation worked correctly we should have 0 Srpms

**Comment by jlsherrill** [comment]
@parthaa thats true for all unit types.  We don't clean up our db at any point, however its not visible to the user at all.  Feel free to file an issue :)

**Comment by parthaa** [comment]
APJ

> @parthaa thats true for all unit types. We don't clean up our db at any point, however its not visible to the user at all. Feel free to file an issue :)

Filed http://projects.theforeman.org/issues/21414

**Comment by ehelms** [comment]
Does this mean I can actually sync an SRPM repository now? And view the contents like I would a normal RPM repository?

**Comment by jlsherrill** [comment]
@ehelms you've been able to sync srpm repos for years (since katello 1.0 iirc), however the following were never supported:

1) adding to a content view, publishing, and having that content show up in the view
2) listing srpm content
3) removing srpm content

This pr fixes 1), but not 2) or 3)

**Comment by ehelms** [comment]
Any plans for #2 or #3? If I learned anything from files, those last two are almost more useful than content view support.

**Comment by jlsherrill** [comment]
This is the first user reported SRPM request i've seen in ~5 years, thus it seems fairly low priority.  I welcome 2) and 3) being done, but i'm not jumping at the seams to implement them given everything else on the table *cough* debian *cough*

**Comment by ehelms** [comment]
If you count me, its two reports in 5 years as I've wanted the ability to upload and manage SRPMs in a repository for dogfooding and CI/CD :) I do think being able to sync or upload to a repository and not see whats in it makes it feel like a black box.

**Comment by jlsherrill** [comment]
@ehelms i agree, feel free to file an issue and feel free to get it prioritized.  My goal here was not to fix *all the things* with srpm.

**Comment by bbuckingham** [comment]
[test]

**Comment by jlsherrill** [comment]
[test default]

**Comment by jlsherrill** [comment]
[test default]

**Comment by jlsherrill** [comment]
[test default]

**Pull request merged**